configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java

```diff
@@ -594,22 +595,16 @@ public HttpResponse fileDistributionStatus(ApplicationId applicationId, Duration
         return fileDistributionStatus.status(getApplication(applicationId), timeout);
     }
 
-    public List<String> deleteUnusedFileDistributionReferences(File fileReferencesPath, Duration keepFileReferencesDuration) {
-        if (!fileReferencesPath.isDirectory()) throw new RuntimeException(fileReferencesPath + " is not a directory");
-
+    public List<String> deleteUnusedFileDistributionReferences(FileDirectory fileDirectory, Duration keepFileReferencesDuration) {
         Set<String> fileReferencesInUse = getFileReferencesInUse();
         log.log(Level.FINE, () -> "File references in use : " + fileReferencesInUse);
         Instant instant = clock.instant().minus(keepFileReferencesDuration);
         log.log(Level.FINE, () -> "Remove unused file references last modified before " + instant);
 
-        List<String> fileReferencesToDelete = sortedUnusedFileReferences(fileReferencesPath, fileReferencesInUse, instant);
+        List<String> fileReferencesToDelete = sortedUnusedFileReferences(fileDirectory.getRoot(), fileReferencesInUse, instant);
         if (fileReferencesToDelete.size() > 0) {
             log.log(Level.FINE, () -> "Will delete file references not in use: " + fileReferencesToDelete);
-            fileReferencesToDelete.forEach(fileReference -> {
-                File file = new File(fileReferencesPath, fileReference);
-                if ( ! IOUtils.recursiveDeleteDir(file))
-                    log.log(Level.WARNING, "Could not delete " + file.getAbsolutePath());
-            });
+            fileReferencesToDelete.forEach(fileReference -> fileDirectory.delete(new FileReference(fileReference)));
```

Comment:
Won't you need to hold the lock while evaluating whether or not to delete the file, too?
The simplest is perhaps to pass a condition to evaluate inside the delete method, and re-run it while holding the lock there?

Reply:
Yeah, that's right, working on it.